### ISSUE TITLE
ItemPositionsNotifier should be visible outside of the package.

### ISSUE THREAD

**Issue by marcglasberg** [google/flutter.widgets#92]
Your class `ItemPositionsListener` is a factory that creates a `ItemPositionsNotifier`.

`ItemPositionsListener` has only a getter for `itemPositions`, while `ItemPositionsNotifier` has both a getter and a setter.

But `ItemPositionsNotifier` is not visible outside of the package, so I can't extend it without a warning.

I need to extend it to test and add some functionality, so, could you please make `ItemPositionsNotifier` visible?

**Comment by tarobins** [comment]
I'm not super keen on exposing `ItemPositionsNotifier`. What functionality do you want to add?  Maybe there's another way to go about it.

**Comment by marcglasberg** [comment]
@tarobins It's just generally a good practice to let people extend things. But I'm going to close this. This issue is not very important, because you can import ItemPositionsNotifier anyway, even if you get a warning.

**Comment by tarobins** [comment]
If you want to add methods to `ItemPositionsListener`, you can do something like

```
extension FirstItem on ItemPositionsListener {
  int get first => itemPositions.value
      .where((ItemPosition position) => position.itemTrailingEdge > 0)
      .reduce((ItemPosition min, ItemPosition position) =>
          position.itemTrailingEdge < min.itemTrailingEdge ? position : min)
      .index;
}
```

and import whatever file contains that extension.  

then in your code use `itemPositionsListener.first`

**Comment by jamesderlin** [comment]
> @tarobins It's just generally a good practice to let people extend things. But I'm going to close this. This issue is not very important, because you can import ItemPositionsNotifier anyway, even if you get a warning.

I disagree that it is generally good practice.  For the same reasons that a class might have private members, a library might want private implementation classes.  Always making things extensible can be a hindrance to future development since breaking changes would become a greater concern.

Be aware that if you directly import internal files that you are at greater risk for being broken if the implementation changes.

**Comment by marcglasberg** [comment]
@jamesderlin I am saying it's a good practice to let people extend things somehow. I am NOT saying it's good practice that implementation classes should be public.

A library might want private implementation classes, sure, but then implementation classes shouldn't leak to the public API. It's actually leaking here, since in this case there is no way to extend it without ItemPositionsNotifier.

The right way to fix this would be not to make ItemPositionsNotifier public as I suggested, but, instead, make ItemPositionsNotifier implement a public interface, and then use that public interface.

That's actually why I asked it to be made public: so that I don't risk being broken if the implementation changes.

In any case, I was just being too perfectionist here. There are more important issues than this one, and I closed this one already.

I read all issues, and my opinion at the moment is that the most important issues to fix are the ones raising exceptions, and the ones that make it impossible for you to do important/common things. These: https://github.com/google/flutter.widgets/issues/38, and https://github.com/google/flutter.widgets/issues/99, and https://github.com/google/flutter.widgets/issues/95.
